drivers/console/uart_console.c

```diff
@@ -596,6 +599,15 @@ static int uart_console_init(struct device *arg)
 	uart_console_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
 
 #if defined(CONFIG_USB_UART_CONSOLE) && defined(CONFIG_USB_UART_DTR_WAIT)
+#ifndef CONFIG_USB_DEVICE_AUTO_ENABLE
```

Comment:
To be honest, I would not expect `usb_enable()` here. We probably have to add a mutex to `usb_enable()`.

```diff
diff --git a/subsys/usb/usb_device.c b/subsys/usb/usb_device.c
index 96727c5af3..339cf9521d 100644
--- a/subsys/usb/usb_device.c
+++ b/subsys/usb/usb_device.c
@@ -110,6 +110,8 @@ LOG_MODULE_REGISTER(usb_device);
 extern struct usb_cfg_data __usb_data_start[];
 extern struct usb_cfg_data __usb_data_end[];
 
+K_MUTEX_DEFINE(usb_lock);
+
 struct usb_transfer_data {
        /** endpoint associated to the transfer */
        u8_t ep;
@@ -1542,10 +1544,12 @@ int usb_enable(void)
                return 0;
        }
 
+       k_mutex_lock(&usb_lock, K_FOREVER);
+
        /* Enable VBUS if needed */
        ret = usb_vbus_set(true);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        usb_register_status_callback(forward_status_cb);
@@ -1553,7 +1557,7 @@ int usb_enable(void)
 
        ret = usb_dc_attach();
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Configure control EP */
@@ -1563,32 +1567,32 @@ int usb_enable(void)
        ep0_cfg.ep_addr = USB_CONTROL_OUT_EP0;
        ret = usb_dc_ep_configure(&ep0_cfg);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ep0_cfg.ep_addr = USB_CONTROL_IN_EP0;
        ret = usb_dc_ep_configure(&ep0_cfg);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
        /* Register endpoint 0 handlers*/
        ret = usb_dc_ep_set_callback(USB_CONTROL_OUT_EP0,
                                     usb_handle_control_transfer);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ret = usb_dc_ep_set_callback(USB_CONTROL_IN_EP0,
                                     usb_handle_control_transfer);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Register endpoint handlers*/
        ret = composite_setup_ep_cb();
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        /* Init transfer slots */
@@ -1600,17 +1604,20 @@ int usb_enable(void)
        /* Enable control EP */
        ret = usb_dc_ep_enable(USB_CONTROL_OUT_EP0);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        ret = usb_dc_ep_enable(USB_CONTROL_IN_EP0);
        if (ret < 0) {
-               return ret;
+               goto usb_enable_error;
        }
 
        usb_dev.enabled = true;
 
-       return 0;
+usb_enable_error:
+       k_mutex_unlock(&usb_lock);
+
+       return ret;
 }
 
 /*

```

Reply:
I actually do not get what's the point for that. By the Kconfig option and #ifndefs I assume the usb_enable will be called once. Is mutex really necessary in this case?

Reply:
For example a application has 3 interfaces: 2 x CDC ACM, one of them for console; 1 x MSC

CONFIG_USB_COMPOSITE_DEVICE=y
CONFIG_USB_CDC_ACM=y
CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
...
CONFIG_USB_UART_CONSOLE=y
CONFIG_USB_DEVICE_AUTO_ENABLE=n
CONFIG_USB_MASS_STORAGE=y

Is it safe that usb_enable() can not be interrupted and than be called by another thread?

Reply:
Okay, now I see your point and ofc You are right. I will introduce mutex here.